### PR TITLE
#1914 Use default values for fields of input types when default argument value is provided

### PR DESCRIPTION
**Pull request by priyaaggarwal24**
Adds support for honoring the default values declared in input type definition ([Issue](https://github.com/graphql-java/graphql-java/issues/1914))

### Example 1
Schema:
```
type Query {
    sayHello(arg: Arg! = {}): String
}

input Arg {
    foo: String = "Bar"
}
```
Query:
```
query {
    sayHello
}
```
arg value before change:
`{}`

arg value after change:
`{foo: "Bar"}`

### Example 2
Schema:
```
type Query {
    sayHello(arg: Arg! = {field1: "F1ValOverride"}): String
}

input Arg {
    field1: String = "F1V"
    field2: String = "F2V"
}
```
Query:
```
query {
    sayHello
}
```
arg value before change:
`{field1: "F1ValOverride"}`

arg value after change:
`{field1: "F1ValOverride", field2: "F2V"}`

### CODE REVIEW

File: src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java [hunk 1]
```diff
@@ -292,8 +293,13 @@ Object buildArrayValue(GraphQLType requiredType, ArrayValue arrayValue) {
 
     Object buildObjectValue(ObjectValue defaultValue, GraphQLInputObjectType objectType) {
         Map<String, Object> map = new LinkedHashMap<>();
-        defaultValue.getObjectFields().forEach(of -> map.put(of.getName(),
-                buildValue(of.getValue(), objectType.getField(of.getName()).getType())));
+        objectType.getFieldDefinitions().forEach(
+                f -> {
+                    final Value<?> fieldValueFromDefaultObjectValue = defaultValue.getObjectFields().stream()
```
Comment (priyaaggarwal24) [inline src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java:298]: Performance of this statement can be improved by keeping a map of Object field's name and the field itself in ObjectValue class.

File: src/test/groovy/graphql/introspection/IntrospectionResultToSchemaTest.groovy [hunk 2]
```diff
@@ -590,7 +590,7 @@ input CharacterInput {
 }
 
 type Query {
-  outputField(inputArg: InputType = {age : 666, name : "nameViaArg"}, inputBoolean: Boolean = true, inputInt: Int = 1, inputString: String = "viaArgString"): OutputType
+  outputField(inputArg: InputType = {age : 666, complex : {boolean : true, int : 666, string : "string"}, name : "nameViaArg", rocks : true}, inputBoolean: Boolean = true, inputInt: Int = 1, inputString: String = "viaArgString"): OutputType
```
Comment (priyaaggarwal24) [inline src/test/groovy/graphql/introspection/IntrospectionResultToSchemaTest.groovy:593]: Introspection of schema now considers the default values defined in Input types.

**Comment by bbakerman** [comment]
Given:

```
sayHello(arg: Arg! = {foo : null}): String
```

what's the value?  I expect `null` not the default in this case

I havent look in the code yet or the tests but can you answer this please either way

**Comment by priyaaggarwal24** [comment]
I, too, expect null value, @bbakerman Let me add a test for this use case as well.

File: src/test/groovy/graphql/Issue1914.groovy [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            return arg.get("field1") + " & " + arg.get("field2")
+        } as DataFetcher
+        def graphQL = TestUtil.graphQL(spec, ["Query": ["sayHello": df]]).build()
+
+        when:
+        def result = graphQL.execute('{sayHello}')
+
+        then:
+        result.errors.isEmpty()
+        result.data == [sayHello: "F1ValOverride & F2V"]
```
Comment (bbakerman) [inline src/test/groovy/graphql/Issue1914.groovy:101]: Can we get another test in place for when the inner field is named BUT the value is null

**Review by bbakerman — CHANGES_REQUESTED** [review]

File: src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java (hunk 1 above)
Comment (priyaaggarwal24) [reply src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java:298]: ok, thanks

File: src/test/groovy/graphql/Issue1914.groovy (hunk 3 above)
Comment (priyaaggarwal24) [reply src/test/groovy/graphql/Issue1914.groovy:101]: do you something like 
"default values in input objects are overridden when null value is provided in input"?

or do you mean the default value provided for a field in input type is null?

Comment (priyaaggarwal24) [reply src/test/groovy/graphql/Issue1914.groovy:101]: @bbakerman Added a test where one of the fields in input type have null value (below this test)

File: src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java [hunk 4]
```diff
@@ -292,8 +293,13 @@ Object buildArrayValue(GraphQLType requiredType, ArrayValue arrayValue) {
...
+                            .filter(dvf -> dvf.getName().equals(f.getName())).map(ObjectField::getValue).findFirst().orElse(null);
+                    map.put(f.getName(), fieldValueFromDefaultObjectValue != null ? buildValue(fieldValueFromDefaultObjectValue, f.getType()): f.getDefaultValue());
+                }
+        );
         return map;
```
[the remark this answers is not in the document]
Comment (priyaaggarwal24) [reply src/main/java/graphql/schema/idl/SchemaGeneratorHelper.java:302]: @bbakerman Done. Refactored the long statement into a method.
